src/adapter/src/catalog.rs

```diff
@@ -158,6 +160,82 @@ pub struct CatalogPlans {
 }
 
 impl Catalog {
+    /// Initializess the `storage_controller` to understand all shards that
+    /// `self` expects to exist.
+    ///
+    /// Note that this must be done before creating/rendering collections
+    /// because the storage controller might not be aware of new system
+    /// collections created between versions.
+    async fn initialize_storage_controller_state(
+        &mut self,
+        storage_controller: &mut dyn StorageController<Timestamp = mz_repr::Timestamp>,
+        builtin_migration_metadata: BuiltinMigrationMetadata,
+    ) -> Result<(), mz_catalog::durable::CatalogError> {
+        let collections = self
+            .entries()
+            .filter(|entry| entry.item().is_storage_collection())
+            .map(|entry| entry.id())
+            .collect();
+
+        let mut state = self.state.clone();
```

Comment:
Left a comment, but I'm doing a little bit of a song and dance here just to ensure that the catalog state doesn't get "modified" until we know that everything succeeds, i.e. we don't mutate state so that everything remains unchanged if we return an error.